### DISCUSSION TITLE
Browser/GPU crash on iOS/Android

### DISCUSSION THREAD

**Discussion by hybridherbst** [google/model-viewer#3639] [Q&A]
### Description

The attached file crashes Chrome on Android and Chrome/Safari on iOS. It might be because that file simply is too big (it contains very very simple geometry, but a lot of transparent materials); still I'm not sure if it should result in a crash.

#### Live Demo
[TransparencyTest.zip](https://github.com/google/model-viewer/files/6137932/TransparencyTest.zip)
Open this in https://modelviewer.dev/editor and use the "deploy mobile" button to load in Android/iOS.

### Version

- model-viewer: v1.6.0

### Browser Affected

- [X] Chrome
- [ ] Edge
- [ ] Firefox
- [ ] IE
- [X] Safari

### OS

- [X] Android
- [X] iOS
- [ ] Linux
- [ ] MacOS
- [ ] Windows

### AR

Interestingly, this file also crashes https://arvr.google.com/scene-viewer-preview and also SceneViewer (tried manually invoking the intent). The converted USDZ version works in AR QuickLook on iOS, albeit slow.

- [X] WebXR
- [X] SceneViewer
- [ ] QuickLook

**Comment by elalish** [comment]
I'm guessing you're just running out of GPU RAM. @kenrussell Does this sound like a Chrome bug (given the crash) or just something we don't really expect to work?

**Comment by hybridherbst** [comment]
Wouldn't it be a bug if that runs out of GPU RAM? This is a tiny (100KB) file...

**Comment by elalish** [comment]
Oh, when you said the file was too big, I assumed it had huge textures. Interesting.

**Comment by kenrussell** [comment]
The first time I tried loading this model from inside Google Lens (which uses a WebView) it crashed Chrome's GPU process, so the model failed to load. Then opening the page in full Chrome, it loaded successfully, and also loaded successfully upon reloading.

My best guess is that this model is causing a lot of different material instances to be created at run time, each of which causes an expensive shader compilation and program link step to occur. Ultimately this is causing the main thread of Chrome's GPU process to hang for too long and its watchdog to kill it. I haven't debugged into Three.js to verify this guess.

If this is correct, it shouldn't really happen; the GPU process's main thread is supposed to notify the watchdog after each shader is compiled.

My best recommendations are:

- If these materials all differ only by a small amount, like color and transparency amount, they should all be using the same program, just with different uniform values or per-vertex attributes. Per-vertex attributes should be preferred so that more geometry can be batched together in a single draw call.
- model-viewer should instantiate its materials one at a time, especially on mobile devices, in a requestAnimationFrame loop so that the application stays responsive while the model's being loaded.
- Ultimately, the KHR_parallel_shader_compile extension should be used so that shader compilation and program linking happens in parallel, though this will take a little while to make it to Chrome on Android. This extension is already on Windows and Linux, and coming soon to macOS, so integration of it in Three.js can happen now, if it hasn't already.

Happy to provide more advice, but we would need a smaller test case in order to investigate this further on the browser side. Something closer to the tests in https://github.com/KhronosGroup/WebGL/tree/master/sdk/tests/conformance/rendering .

CC @mrdoob @jdarpinian
